Dashboard + Checklist + Cohesiveness

claudiulodro opened the pull request:
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request, and the reason for such changes. -->

Closes #56 .

Sorry for YAGPR (Yet Another Giganto Pull Request 😄). The components are somewhat woven together, and I didn't feel that the architecture could be properly evaluated or communicated if it was split into a bunch of smaller PRs.

This pull request implements dashboard and checklist handling and ties everything together into one (mostly) functional application.

Architecture:
- The Dashboard is a stand-alone area that can render anything into a card (checklists, links, etc.). Under the hood, it's just a 1-screen wizard. See `assets/wizards/dashboard` and `includes/wizards/class-dashboard.php`.
- Checklists are managed and rendered in a centralized way, so we don't need to make a bunch of different independent checklist handlers. Checklists are defined and routed in  `includes/class-checklists.php`. Checklists are rendered in `assets/wizards/checklist`.
- Wizards contain all of the info in their classes (name, description, etc.). They are managed and initialized in `includes/class-wizards.php`. The `Wizards` class provides a central location for retrieving wizards and their info.

Screenshots in details:
<details>

<img width="1449" alt="Screen Shot 2019-06-10 at 11 20 32 AM" src="https://user-images.githubusercontent.com/7317227/59218291-c2ab8a80-8b74-11e9-8e90-57b266323125.png">
<img width="205" alt="Screen Shot 2019-06-10 at 11 20 40 AM" src="https://user-images.githubusercontent.com/7317227/59218292-c2ab8a80-8b74-11e9-947e-64a49b2df311.png">
<img width="1169" alt="Screen Shot 2019-06-10 at 11 21 00 AM" src="https://user-images.githubusercontent.com/7317227/59218294-c3442100-8b74-11e9-8a68-73bfa535468c.png">
<img width="1119" alt="Screen Shot 2019-06-10 at 11 21 14 AM" src="https://user-images.githubusercontent.com/7317227/59218295-c3442100-8b74-11e9-930b-e239d6317acf.png">

</details>

### How to test the changes in this Pull Request:

1. `npm ci; npm run clean; npm run build:webpack`
2. Go to WP Admin > Newspack. Observe most wizards don't have links in the submenu any more.
3. On the dashboard, observe things look good, with a mix of statuses on the cards.
4. Click the "Demo Checklist" or "Memberships" cards. Observe you are taken to the appropriate checklist.
5. Click the "Skip" in the checklist to skip an item or "Do it" to go to the appropriate wizard. Click "I'm done setting up" to return to the dashboard.

The main thing not done in this PR is making the checklist item completion permanent; that'll be done in a future PR.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

<!-- Mark completed items with an [x] -->

claudiulodro commented:
Probably don't want to merge this until #89 is ready, so we're sure it all works together nicely.

philipjohn commented on includes/wizards/class-subscriptions-wizard.php:
```diff
@@ -49,6 +42,33 @@ public function __construct() {
+	 * @return string The wizard name.
+	 */
+	public function get_name() {
+		return __( 'Subscriptions', 'newspack' );
```
We should always use the escaping alternatives for localisation functions where possible to ensure that translations don't introduce any unsafe output. E.g. `esc_html__()`

philipjohn commented on includes/wizards/class-wizard.php:
```diff
@@ -87,4 +91,58 @@ public function api_permissions_check( $request ) {
+	 * @return string
+	 */
+	public function get_url() {
+		return admin_url( 'admin.php?page=' . $this->slug );
```
Let's make sure this is safe by wrapping it in `esc_url()` too

claudiulodro commented on includes/class-checklists.php:
```diff
@@ -0,0 +1,224 @@
+	 * Register and add the page that the checklists will live on.
+	 */
+	public static function add_page() {
+		add_submenu_page( 'newspack', 'Checklist', 'Checklist', 'manage_options', 'newspack-checklist', 'Newspack\Checklists::render_checklist' );
```
It still needs to be an admin page; we just don't want a link to it to show up in the admin sidebar. Does that answer your question? I'm not super sure exactly what you're saying here. :)

philipjohn replied:
Oh, does WP need use to call `add_submenu_page()` in order for the page to be accessible at all, even if we hide it from the menu? Perhaps that's the nuance I was missing in my understanding.

claudiulodro replied:
Yes, exactly! There's all sorts of permissions handling, routing, etc. that `add_submenu_page` handles. To test you can try going to a non-existent page e.g. `http://local.newspack.test/wp-admin/admin.php?page=blah` and see you're prevented from accessing it.

claudiulodro commented on includes/class-checklists.php:
```diff
@@ -0,0 +1,224 @@
+			'newspack-checklist',
+			Newspack::plugin_url() . '/assets/dist/checklist.js',
+			[ 'wp-components' ],
+			filemtime( dirname( NEWSPACK_PLUGIN_FILE ) . '/assets/dist/checklist.js' ),
```
For sure, I mostly didn't want the overhead of managing versions and having to make sure caches are busted while we're developing the plugin. Is it cool with you if we wait until the plugin is closer to release before making that change?

philipjohn replied:
Definitely! I've created #99 so we don't forget :)

claudiulodro commented:
Thanks for the feedback, @philipjohn! I'll get your requested changes made before merging.

> One thing I did encounter, which may be entirely unrelated to this change was that going through the subscriptions wizard I noticed that once I'd added a subscription there was no way to get back to the checklist without going back through the dashboard.

Yep; that needs to be added to any existing wizards. It's definitely part of the original designs, and was omitted because there was nowhere for the link to point to until now. I've created an issue around this at https://github.com/Automattic/newspack-plugin/issues/98

claudiulodro commented:
Requested changes made (See commit log if interested in specifics):
- Removed unused `init_hooks`.
- Added some php comments around the localized JS variables.
- Added escaping where needed.
